### PR TITLE
TST: Clean up tests of DataFrame.sort_{index,values}

### PR DESCRIPTION
**Pull request by IamJeffG**
Before taking a stab at #10806, I looked at the relevant tests and they didn't make much sense to me.

This commit fixes:
- Some expected results were obtained by running code identical to what's being tested. (i.e. test could never fail)
- Removed duplicate assertions testing the same functionality both in `test_sort_index` and in `test_sort_values`.
- Switched the names `test_sort_index` and `test_sort_values` based on the (remaining) assertions being tested in their bodies.

Still confusing, unfixed:
- The `DataFrame.sort_index` docstring doesn't specify what happens when `level` is unspecified and `sort_remaining=False`. However there is a legacy test using this case: it's not clear to me what result or behavior is being tested.

First time pandas contributor, so I'd appreciate a good critical review!
- [X] tests pass
- [X] passes `git diff upstream/master | flake8 --diff`
- [x] ensure default options in docstrings

### CODE REVIEW

**Comment by jreback** [comment]
ok, pls factor out ``tests/series/tests_sorting.py`` while you are at it. (in test_analytics).
Ideally make sure we have test coverage for all of the options for sort_values/index.

I think these tests were this way originally because before 0.17.0 sort did *both* sort index/values in some cases.

looks good.

**Comment by jorisvandenbossche** [comment]
> The DataFrame.sort_index docstring doesn't specify what happens when level is unspecified and sort_remaining=False. However there is a legacy test using this case: it's not clear to me what result or behavior is being tested.

I think should do nothing. When you do not specify `level`, I think this means that you want to sort by all levels, and in this way `sort_remaining` (either true or false) has no effect anymore.

I think this is also what the test does? (the same result for both `sort_remaining` is True or False)

**Comment by jreback** [comment]
yes, the doc-string should have the default for ``sort_remaining`` and pls check the other params, a stock string should always list the default (even if ``None``)

**Comment by IamJeffG** [comment]
@jorisvandenbossche +1 to your interpretation. I might clarify the code comment in the test to explain.

@jreback Sure, I'll fix up the docstrings and also look at `tests/series/tests_sorting.py`

**Comment by IamJeffG** [comment]
I could use guidance on a few small details:

1. Numpy Docstring Standard says just to mark optional kwargs as "optional", not to re-state their default value in the docstring. @jreback do you really want the defaults listed?
2. Is a whatsnew entry necessary for this change, which merely refactors tests without touching the core source?
3. The `sort_values` docstring is shared (`core/generic.py`) and therefore includes parameter "by" even for Series where it's not accepted. What's the right way to handle this discrepancy? Is it necessary to "un-share" the docstring?

**Comment by jreback** [comment]
> Numpy Docstring Standard says just to mark optional kwargs as "optional", not to re-state their default value in the docstring. @jreback do you really want the defaults listed?

yes, we do this for most other things, when you are reading a doc-string, its annoying to refer back to the signature. I am not sure we are very consistent in this though. @jorisvandenbossche 

> Is a whatsnew entry necessary for this change, which merely refactors tests without touching the core source?

In this case no, if its a 'big' test refactor then we usually want to note it

> The sort_values docstring is shared (core/generic.py) and therefore includes parameter "by" even for Series where it's not accepted. What's the right way to handle this discrepancy? Is it necessary to "un-share" the docstring?

Right way is to share, but add in ``by`` via string replacement specifically for ``DataFrame`` (to avoid repeating things more than necessary)

**Comment by jorisvandenbossche** [comment]
> > Numpy Docstring Standard says just to mark optional kwargs as "optional", not to re-state their default value in the docstring. @jreback do you really want the defaults listed?

> yes, we do this for most other things, when you are reading a doc-string, its annoying to refer back to the signature. I am not sure we are very consistent in this though. @jorisvandenbossche 

Well, the question if how you define 'optional'. In principle are almost all keyword arguments optional (in the sense that you do not have to pass them explicitly), but in many cases they have a meaningful default value. And then it is indeed informative to note that as @jreback says.

**Comment by IamJeffG** [comment]
> add in `by` via string replacement specifically for `DataFrame`

@jreback Do you mean like this? https://github.com/pydata/pandas/pull/13496/commits/a12cea532c8db68641173150e6b984d359395ea3

(However now I'm wondering if "by" will after all become used for Series, `axis=1`, per #10806.)

Besides `by`, also `Series.sort_index` does not accept the kwargs `kind` and `na_position`. Until they get implemented I feel I should do the same sort of string replacement, although this begins to get unweildy.

**Comment by jorisvandenbossche** [comment]
> Do you mean like this?

Yes, that looks good

> (However now I'm wondering if "by" will after all become used for Series, axis=1, per #10806.)

No, I think that will also only be for dataframes (Series has no axis=1)

> Besides by, also Series.sort_index does not accept the kwargs kind and na_position. Until they get implemented I feel I should do the same sort of string replacement, although this begins to get unweildy.

Indeed, you can change that as well. 
But, it would not be too difficult to add those parameters to Series.sort_index (it's just changing the numpy argsort call with the internal wrapper `_nargsort` which adds those additional keywords. But if you want to do this, maybe keep that for a separate PR.

**Comment by jreback** [comment]
ping when ready with this

**Comment by IamJeffG** [comment]
> Series.sort_index does not accept the kwargs kind and na_position

Opened #13589 for these missing features.

> make sure we have test coverage for all of the options for sort_values/index

My report on test coverage:
* `sort_values`'s tests for multiple by's were naive. I've improved this by adding duplicate values to the DataFrame being tested.
* `sort_index` was missing a test for multi-level `Series.sort_index(sort_remaining)`. I've added it.

I think this is about ready for merge, @jreback

File: pandas/tests/frame/test_sorting.py [hunk 1]
```diff
@@ -21,75 +21,63 @@ class TestDataFrameSorting(tm.TestCase, TestData):
 
     _multiprocess_can_split_ = True
 
-    def test_sort_values(self):
-        # API for 9816
-
-        # sort_index
+    def test_sort_index(self):
         frame = DataFrame(np.arange(16).reshape(4, 4), index=[1, 2, 3, 4],
```
Comment (jreback) [inline pandas/tests/frame/test_sorting.py]: add this issue number (so future readers can reference)

File: pandas/tests/frame/test_sorting.py [hunk 2]
```diff
@@ -21,75 +21,63 @@ class TestDataFrameSorting(tm.TestCase, TestData):
...
-        assert_frame_equal(result, expected)
+        assert_frame_equal(result, frame)
+
+        result = unordered.sort_index(axis=1, ascending=False)
+        expected = frame.ix[:, ::-1]
         assert_frame_equal(result, expected)
 
-        # sortlevel
-        mi = MultiIndex.from_tuples([[1, 1, 3], [1, 1, 1]], names=list('ABC'))
+        # sort rows by specified level of multi-index
```
Comment (jreback) [inline pandas/tests/frame/test_sorting.py]: I would make multi-index part of a 2nd test (e.g. separate long tests into multiple ones)

File: pandas/tests/series/test_sorting.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
+# coding=utf-8
+
```
Comment (jreback) [inline pandas/tests/series/test_sorting.py]: good idea, new file

**Comment by jreback** [comment]
very minor comments. pls post the new doc-strings, and ping on green.

**Comment by jreback** [comment]
@IamJeffG this looks good, except the doc-string on DataFrame.sort_values is looking a bit funny. The by is not indented properly.

**Comment by IamJeffG** [comment]
@jreback see if my newest commit fixes this. I could verify the indentation problem using string replacement in the REPL, but I was unable to get docs to successfully build, to verify my fix end-to-end.

**Comment by jorisvandenbossche** [comment]
@IamJeffG Looks good now! (BTW, you can also just look at the docstring in the interactive console (`pd.DataFrame.sort_values?` in ipython) to check that it looks OK)

**Comment by IamJeffG** [comment]
Thanks for the good tip @jorisvandenbossche!

**Comment by IamJeffG** [comment]
@jreback We're green.

**Comment by jorisvandenbossche** [comment]
@IamJeffG Thanks a lot!

**Comment by jorisvandenbossche** [comment]
Interested in taking on https://github.com/pydata/pandas/issues/13589 as well?

**Comment by IamJeffG** [comment]
Perhaps. Let me start with #10806, which was my initial goal.

**Pull request merged**